p1c2u/openapi-core#296: Implements OpenAPI 3.1 validator

nezhar opened the pull request:
Support for [draft 2020-12](https://json-schema.org/draft/2020-12/json-schema-validation.html) has been added to [jsonschema](https://github.com/Julian/jsonschema) with https://github.com/Julian/jsonschema/pull/817

Since there is no release at the moment, this is only a draft PR as it installs jsonschema directly from github.

The **OAS31Validator** has been added by extending the **Draft202012Validator**. The validator can be used with the `draft202012_format_checker` as seen in the tests.

Python `2.7` and `3.5` have been removed from the test matrix as they are end of life.

nezhar commented:
@p1c2u I would appreciate if you have any sugestions or feedback for the implementation.

I will build the support of OpenAPI 3.1 for https://github.com/p1c2u/openapi-spec-validator on top of this PR and I will use the tests provided in https://github.com/OAI/OpenAPI-Specification/tree/main/tests/v3.1

p1c2u commented on requirements.txt:
```diff
@@ -1,5 +1,5 @@
 isodate
-jsonschema
+git+https://github.com/Julian/jsonschema@main#egg=jsonschema[format]
```
we need to wait for alpha version to be released with the change
https://github.com/Julian/jsonschema/commit/1a7a935fd3e4dc11089b0a9f6ac247d66dc6b358

nezhar replied:
sure, I used the `main` branch for now in order to be able to start the implementation :slightly_smiling_face:

nezhar replied:
Switched to [4.0.0a6](https://pypi.org/project/jsonschema/4.0.0a6/)

Also dropped Python 3.6 support as in jsonschema

ngnpope replied:
@nezhar It looks like a stable release ([v4.0.1](https://pypi.org/project/jsonschema/4.0.1/)) is now available.

p1c2u reviewed (commented):
We will also need:
* coverage of definitions extended by the OAS
* update of README with versions

nezhar commented:
> We will also need:
> 
> * coverage of definitions extended by the OAS
> * update of README with versions

README has been updated. I also made some changes to the validators, but I'm not sure which definitions are meant to be covered with this library. Are all fixed fields required such as the table [here](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#fixed-fields) or ist this out of scope?

p1c2u commented:
Sorry I wasn't clear enough. I meant this one:

> The following properties are taken from the JSON Schema specification but their definitions have been extended by the OAS:
> 
> description - CommonMark syntax MAY be used for rich text representation.
format - See Data Type Formats for further details. While relying on JSON Schema's defined formats, the OAS offers a few additional predefined formats.

https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#schema-object

and fixed fields [here](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#fixed-fields-20)

nezhar commented:
> Sorry I wasn't clear enough. I meant this one:
> 
> > The following properties are taken from the JSON Schema specification but their definitions have been extended by the OAS:
> > description - CommonMark syntax MAY be used for rich text representation.
> > format - See Data Type Formats for further details. While relying on JSON Schema's defined formats, the OAS offers a few additional predefined formats.
> 
> https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#schema-object
> 
> and fixed fields [here](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#fixed-fields-20)

I added the missing fields. Should I also create new validators for the fields? I would implement them as follows:

- description - String
- discriminator - covers the [discriminator](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#discriminatorObject) definition
- xml - covers the [XML Object](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#xmlObject) definition
- externalDocs - covers the [external documents](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#externalDocumentationObject) definition
- example - covers [example object](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#exampleObject) definition

warmwaffles commented:
@nezhar and @p1c2u any updates for this?

nezhar commented:
> @nezhar and @p1c2u any updates for this?

Currently this is waiting to be reviewed, you may try to install the fork meanwhile :slightly_smiling_face:

tschmidtb51 commented:
@p1c2u Could you please add your review?

p1c2u reviewed (approved):
LGTM just needs to be rebased

tschmidtb51 commented:
@nezhar please rebase.

nezhar commented:
@p1c2u @tschmidtb51 PR is rebased :slightly_smiling_face:

p1c2u commented:
@nezhar thank you for the contribution

tschmidtb51 commented:
@p1c2u Thanks for merging.
